File: java/src/test/java/site/markhenrick/adventofcode/y2015/Day01Test.java
package site.markhenrick.adventofcode.y2015;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;
import site.markhenrick.adventofcode.common.TestUtil;

import static org.assertj.core.api.AssertionsForClassTypes.assertThat;

class Day01Test {
	private static final String MY_INPUT = TestUtil.getResourceAsString("input/2015/day01.txt");

	@ParameterizedTest
	@CsvSource({
		"(()),		0",
		"(((,		3",
		"(()(()(,	3",
		"))(((((,	3",
		"()),		-1",
		"))(,		-1",
		"))),		-3",
		")())()),	-3",
	})
	void solvePart1Samples(final CharSequence input, final int expected) {
		assertThat(Day01.solvePart1(input)).isEqualTo(expected);
	}

	@Test
	void solvePart1Actual() {
		assertThat(Day01.solvePart1(MY_INPUT)).isEqualTo(280);
	}

	@ParameterizedTest
	@CsvSource({
		"),		1",
		"()()),	5",
	})
	void solvePart2Samples(final CharSequence input, final int expected) {
		assertThat(Day01.solvePart2(input)).isEqualTo(expected);
	}

	@Test
	void solvePart2Actual() {
		assertThat(Day01.solvePart2(MY_INPUT)).isEqualTo(1797);
	}
}

File: kotlin/src/test/kotlin/site/markhenrick/recreational/adventofcode/y2019/Day01Test.kt
package site.markhenrick.recreational.adventofcode.y2019

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.CsvSource
import site.markhenrick.recreational.getChallengeInput

class Day01Test {
    @Test
    fun part1() {
        assertThat(Day01.solvePart1(challengeInput)).isEqualTo(3226822)
    }

    @Test
    fun part2() {
        assertThat(Day01.solvePart2(challengeInput)).isEqualTo(4837367)
    }

    @ParameterizedTest
    @CsvSource(
        "12, 2",
        "14, 2",
        "1969, 654",
        "100756, 33583",
    )
    fun fuelRequired(input: Int, expected: Int) {
        assertThat(Day01.fuelRequired(input)).isEqualTo(expected)
    }

    @ParameterizedTest
    @CsvSource(
        "14, 2",
        "1969, 654+216+70+21+5",
        "100756, 33583+11192+3728+1240+411+135+43+12+2",
    )
    fun fuelSequence(input: Int, expectedRaw: String) {
        val expectedList = expectedRaw.split("+").map { it.toInt()}
        val actualList = Day01.fuelSequence(input).toList()
        assertThat(actualList).isEqualTo(expectedList)
    }

    companion object {
        private val challengeInput = getChallengeInput(2019, 1)
    }
}
